### PR TITLE
feat: define KongUpstreamPolicy CRD

### PR DESCRIPTION
**Pull request by czeslavo**
**What this PR does / why we need it**:

Defines `KongUpstreamPolicy` Policy Attachment API _loosely_ following [GEP-713](https://gateway-api.sigs.k8s.io/geps/gep-713).

**Which issue this PR fixes**:

Part of #3174.

**Special notes for your reviewer**:

Questions to reviewers:

1. Do we expect this CRD to work with Ingress as well? It is doable but may seem a little weird as it implements Gateway API-specific stuff like `PolicyTargetReference`, `Override`, etc.
2. If the answer to 1. is NO then do we:
  2.1. keep `KongIngress.Upstream` for Ingress and add `KongUpstreamPolicy` for GW API *Routes OR 
  2.2. deprecate `KongIngress.Upstream` and create a separate resource specifically for Ingress
3. If the answer to 1. is YES (IMO this would be preferable) then do we:
  3.1. implement all the expected `Conditions` defined by Gateway API also when it's used with Ingress (IMO this could be extracted to a separate issue after KIC 3.0) OR
  3.2. just use the specification (similarly to what we do currently with `KongIngress` and do not populate status at all (IMO this could be a good starting point for KIC 3.0)
3. Does it make sense to add `HostHeader` given Gateway API supports `HTTPHeaderFilter`? (IMO it doesn't)
5. Do we want to rely on `go-kong`'s `Healthcheck` types? (IMO we shouldn't as Kubernetes API definitions should reside in KIC repo, it doesn't make sense to leak Kubernetes API definitions to `go-kong` like it happened for `KongIngress`)

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Comment by pmalek** [comment]
> 1. Do we expect this CRD to work with Ingress as well? It is doable but may seem a little weird as it implements Gateway API-specific stuff like `PolicyTargetReference`, `Override`, etc.

I'd say no. This may be used as a selling point for Gateway API.

> 2. If the answer to 1. is NO then do we:
>    2.1. keep `KongIngress.Upstream` for Ingress and add `KongUpstreamPolicy` for GW API *Routes OR
>    2.2. deprecate `KongIngress.Upstream` and create a separate resource specifically for Ingress

As https://github.com/Kong/kubernetes-ingress-controller/issues/3174 says:

> Some of KongIngress fields can be migrated into annotations, some (that configure upstream behavior) require their own CRD.

I'd lean towards 2.2

> 5. Do we want to rely on `go-kong`'s `Healthcheck` types? (IMO we shouldn't as Kubernetes API definitions should reside in KIC repo, it doesn't make sense to leak Kubernetes API definitions to `go-kong` like it happened for `KongIngress`)

Agreed, let's not use those

File: pkg/apis/configuration/v1/kongupstreampolicy_types.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
+package v1
+
+// REVIEW: do we want to create a separate package for Gateway API extensions?
```
Comment (pmalek) [inline pkg/apis/configuration/v1/kongupstreampolicy_types.go:3]: I'm not sure if we'd like to add a new package but I'd say this should land in `v1` right at the start. `v1beta1` or `v1alpha1` would make much more sense to me.

**Review by rainest — COMMENTED** [review]
In practice, our upstream config only ever needs to attach to Services (which falls on the side of "yes, it can be used with Ingress" indirectly). It only applies to the upstreams generated directly from Services, so there's basically no reason to worry about how it will or will not apply to anything else.

`PolicyTargetReference` is a reasonable generic "this resource does things to this other resource" reference format, though it's explicitly one to one. That breaks with our existing model, where you can reuse configuration across multiple resources.

The [GEP](https://gateway-api.sigs.k8s.io/geps/gep-713) does describe an inheritance model that'd allow you to apply it via a Gateway whose xRoutes then use those Services, but that seems like a rather backwards UX for what this does: you presumably don't want to define, say, whether your Services all use cookie-based or querystring-based session stickiness based on which Gateway (or xRoute) they use.

Given that UX stumbling block, the lack of need to attach this at any level other than Service (a core resource rather than a Gateway API resource), and that the policy attachment stuff remains experimental, I think we should reneg on the original request to use policy attachment for this and stick to annotations. I apologize, I should have done a better review of whether the GWAPI policy attachment proposal was fit for purpose here before saying "yeah, use that".

> Does it make sense to add HostHeader given Gateway API supports HTTPHeaderFilter? (IMO it doesn't)

The new resource will still be usable for Ingress-attached Services, so yeah.

Upstream configuration in general unfortunately covers some things that it shouldn't need to in a Kubernetes world. It functions as an override to directly configure Kong internals, even if some of those are ideally managed elsewhere. Directly mapping it to the actual Kong resource provides a simple interface that maps directly to Kong gateway documentation at the cost of exposing some things you probably shouldn't touch. Healthchecks are in a similar-ish bucket where you should probably leave that up to Pod readiness.

> Do we want to rely on go-kong's Healthcheck types? (IMO we shouldn't as Kubernetes API definitions should reside in KIC repo, it doesn't make sense to leak Kubernetes API definitions to go-kong like it happened for KongIngress)

We probably can so long as we maintain a direct map to the underlying Kong object. It's a bit of extra work to define and maintain overlays. I don't feel strongly one way or the other.

File: pkg/apis/configuration/v1beta1/kongupstreampolicy_types.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+	// Unhealthy configures thresholds and HTTP status codes to mark targets unhealthy for an upstream.
+	Unhealthy *KongUpstreamHealthcheckUnhealthy `json:"unhealthy,omitempty"`
+
+	// HTTPPath is the path to use in GET HTTP request to run as a probe.
+	// +kubebuilder:validation:Pattern=^/.*$
+	HTTPPath *string `json:"http_path,omitempty"`
+
+	// HTTPSSni is the SNI to use in GET HTTPS request to run as a probe.
+	HTTPSSni *string `json:"https_sni,omitempty"`
```
Comment (rainest) [inline pkg/apis/configuration/v1beta1/kongupstreampolicy_types.go:144]:
```suggestion
	HTTPSSNI *string `json:"https_sni,omitempty"`
```

With the caveat that it'll break generated stuff and probably should be done outside GitHub suggest.

This is unfortunately two acronyms stuck together, but I prefer having consistently capitalized acronyms in variables. `HTTPSServerName` or even just `ServerName` also work while looking a bit less silly. Breaking with the JSON name is fine.

**Comment by czeslavo** [comment]
> Given that UX stumbling block, the lack of need to attach this at any level other than Service (a core resource rather than a Gateway API resource), and that the policy attachment stuff remains experimental, I think we should reneg on the original request to use policy attachment for this and stick to annotations. I apologize, I should have done a better review of whether the GWAPI policy attachment proposal was fit for purpose here before saying "yeah, use that".

I agree that annotations could be more user-friendly allowing users to attach the same policy to multiple objects. 

I'm not sure if we should reserve it ONLY for Services though. IMO attaching to an HTTPRoute (or even Ingress) makes sense, because the Upstreams that we generate in Kong are not mapped to Services 1-1. We generate Service-Upstream pairs for `<ingress>.<service>.<port>` and `<route>.<rule-id>` combinations. If we allow attaching only to a Service, we close the ability to more granular control of the underlying Upstreams, because if we attach to a Service, from now on every GW API Route and Ingress using the Service and their generated Upstreams are affected (and if multiple BackendRefs used in *Route, the user has to remember to set the same policy for every Service used in it). I think it may not always be the intention to have the same load-balancing strategy per Kubernetes Service. I imagine someone could have *Routes that use the same Service and some of them care and some do not about the session stickiness, let's say in a situation where a single service exposes endpoints that are session-aware, but some are not. In such a situation it would be useful to be able to attach a policy to a *Route or Ingress and affect all of their rules *only* (it has its pros and cons, e.g. you have to split *Routes to the granularity you want your upstream policy configured, but at least it doesn't close the doors to make that happen).

> Healthchecks are in a similar-ish bucket where you should probably leave that up to Pod readiness.

Yeah, I'd love to say Kubernetes health checks cover all of Kong's capabilities, but they do not (e.g. in Kubernetes you cannot set passive probes). Not sure how much usage of them we have in the wild though (maybe we should consider adding some telemetry to see if it's worth maintaining).

> We probably can so long as we maintain a direct map to the underlying Kong object. It's a bit of extra work to define and maintain overlays. I don't feel strongly one way or the other.

I think it's better to keep the mapping in KIC and at least have control over it. Leaving it in `go-kong` puts us in a situation where areas of Kong and Kubernetes are mixing up in an uncontrolled manner (kubebuilder annotations leaked to go-kong which should not be aware of Kubernetes). We can cover the mapping with `kongintegration`-ish tests to verify that the mapping works as expected quite cheaply I believe.

**Comment by rainest** [comment]
> I'm not sure if we should reserve it ONLY for Services though.

Extracting it from the Service is what we're doing currently:

https://github.com/Kong/kubernetes-ingress-controller/blob/v2.12.0/internal/dataplane/kongstate/upstream.go#L96-L98

https://github.com/Kong/kubernetes-ingress-controller/blob/v2.12.0/internal/dataplane/kongstate/kongstate.go#L226-L241

so all we need for parity with KongIngress's current upstream capabilities is the same.

Allowing it to attach elsewhere would be new functionality, and I wouldn't recommend it unless it's the only option: limiting configuration flexibility (and by extension, the number of user decisions needed before using a feature) until you _need_ it is a good thing. It's less closing doors (they weren't open to begin with) than providing a map that has fewer branching paths that ultimately arrive at the same destination.

If you do want per-route flexibility there are still options within the current model. You can create multiple Services with the same Deployment selector (e.g. a FooService and StatefulFooService, where both have the same endpoints but only the latter uses sticky config) or use an approach that allows for client-side control (e.g. using querystring stickiness and omitting the chosen parameter for requests that can be served by any endpoint).

> in Kubernetes you cannot set passive probes

Yeah, that's the "probably" case in "should probably leave that up to Pod readiness". From an architecture design standpoint I'd say try and stick with a single healthchecking system, since not having to worry about situations where you can have disagreement between healthchecking systems--you don't want the passive checker to have dropped all your endpoints while the Pod healthchecker sees no reason to replace them. If you do need passive on top I'd probably recommend going with https://kuma.io/docs/2.4.x/policies/circuit-breaker/ since it has time-based expiration and guards against marking the entire endpoint set unavailable that the upstream passive checker doesn't.

File: pkg/apis/configuration/v1/kongupstreampolicy_types.go (hunk 1 above)
Comment (czeslavo) [reply pkg/apis/configuration/v1/kongupstreampolicy_types.go]: Moved to `v1beta1`. 👍

**Comment by czeslavo** [comment]
I've updated the PR to introduce minimal version of the policy (no TargetRef, no status conditions). PTAL.

**Review by rainest — APPROVED** [review]

**Pull request merged**